Update keyboard_remote.py

MrMep opened the pull request:
I changed the way the component refers to the keyboard: not by "descriptor", but by name.
The fact is that the udev system doesn't always give a name link in /dev/input/by-id folder and the actual /dev/input/eventX file changes automatically. 
For example, if I had my keyboard on /dev/input/event13 and then it disconnected, if something else connects to the system it will get the first input file available, that is /dev/input/event13. If the keyboard then reconnects, it will get /dev/input/event14, thus breaking the configuration.
Now it searches every time for the right input file.
The problem might be that, in case of ha upgrade, the pre-existing configuration won't work. I think there are some guidelines here, but I am not sure.
But I think it's inevitable: the initial idea to use a /dev/input/by-id/ symbolic link was good, but unfortunately it doesn't seem to work with bluetooth devices.

I haven't updated the documentation yet: I'm waiting for some ok about the change in configuration key names.

**Description:**


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

**Example entry for `configuration.yaml` (if applicable):**
```yaml

```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

MrMep committed:
Update keyboard_remote.py

I added a couple of events: keyboard_remote_connected and keyboard_remote_disconnected, useful to trigger some action (for example: play a sound)
I changed the way the component refers to the keyboard: not by "descriptor", but by name.
The fact is that the udev system doesn't always give a name link in /dev/input/by-id folder and the actual /dev/input/eventX file changes automatically. 
For example, if I had my keyboard on /dev/input/event13 and then it disconnected, if something else connects to the system it will get the first input file available, that is /dev/input/event13. If the keyboard then reconnects, it will get /dev/input/event14, thus breaking the configuration.
Now it searches every time for the right input file.
The problem might be that, in case of ha upgrade, the pre-existing configuration won't work. I thing there are some guidelines here, but I am not sure.
I think it's inevitable: the initial idea to use a /dev/input/by-id/ symbolic link was good, but unfortunately it doesn't seem to work with bluetooth devices.

I haven't updated the documentation yet: I'm waiting for some ok about the change in configuration key names.

armills commented:
At most, device name should be an alternate way to choose a device, and not replace descriptor. Also FYI, you can use udev to create reliable symlinks for specific devices. Here is one for my zwave usb stick. You can probably create a similar rule file to reliably symlink your keyboard.

```bash
# /etc/udev/rules.d/10-zwave-usb.rules
ACTION=="add", ATTRS{idVendor}=="10c4", ATTRS{idProduct}=="ea60", SYMLINK+="zwave"
```

MrMep commented:
I had thought about writing an udev rule, but it seemed to me that it would have made the component even more complex to use, from an end user standpoint.
What's the best way to set up alternate config keys: should device_descriptor and device_name be both optional, and then let setup() decide the matter?

armills commented:
Here's an example of exclusive config options, which would be what you want. I do agree, it could be a nice way to choose, is there anything else in the object that's more of an ID? We should plan for the case where the user is connecting two of the same device.

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/device_tracker/asuswrt.py#L41

MrMep committed:
Update keyboard_remote.py

That should do the trick.
You are right: device names can be duplicated, thus they're not reliable in case of multiple devices. Unfortunately, the only other information available is the physical address, even more complicated.
But in case you have just one keyboard remote of same model, the name works just fine. I'll put it in the documentation which, once the code is approved, I will promptly update.

MrMep committed:
Update keyboard_remote.py

MrMep commented:
That should do the trick.
You are right: device names can be duplicated, thus they're not reliable in case of multiple devices. Unfortunately, the only other information available is the physical address, even more complicated.
But in case you have just one keyboard remote of same model, the name works just fine. I'll put it in the documentation which, once the code is approved, I will promptly update.

armills commented:
So I'm guessing it will show up in the documentation, but how are you finding the name of the keyboard? Is it just from the bluetooth control panel?

I think I'm good with this. Even though names aren't perfect, users can always disambiguate with the device descriptor.

MrMep commented:
There are many ways to get the correct name, but mainly the component will show a list at startup, in case it couldn't connect to the device.
[https://github.com/home-assistant/home-assistant/pull/5535/files#diff-278151e183d8f533f855d1923a4aadb5R101](url)
Is that enough?

armills commented:
Ah, there it is. That's good with me.

armills commented on homeassistant/components/keyboard_remote.py:
```diff
@@ -67,22 +47,18 @@
+            'KeyboardRemote: Error: No device_descriptor\
+             or device_name found.'
+            )
+        return
```
Voluptuous will enforce this, so this check isn't necessary.

MrMep replied:
Actually it doesn't, maybe I did something wrong with the Schema?

armills replied:
Sorry about that, my mistake! It looks like this is still an open issue with voluptuous: https://github.com/alecthomas/voluptuous/issues/115

Just keep this as-is then.

armills commented on homeassistant/components/keyboard_remote.py:
```diff
@@ -106,34 +82,60 @@ def _stop_keyboard_remote(_event):
+            devices = [InputDevice(file_name) for file_name in list_devices()]
+            for device in devices:
+                if self.device_name == device.name:
+                    return device
```
If the loop exits without finding anything, emit a debug log similar to below that the keyboard was not connected. Include the name.

MrMep replied:
Mmh, now that I think of it,  actually this function shouldn't write to the log at all: it gets repeatedly called from the event loop, when the keyboard is not connected.

armills replied:
Good point. We're still getting messages on connection changes. Just delete the other log then.

armills commented on homeassistant/components/keyboard_remote.py:
```diff
@@ -106,34 +82,60 @@ def _stop_keyboard_remote(_event):
-                'KeyboardRemote: keyboard not connected, %s',
-                self.device_descriptor)
-            self.keyboard_connected = False
+                'KeyboardRemote: keyboard connected'
```
Let's still include the keyboard ID in the log messages. In `__init__`, let's include `self.device_id = self.device_name or self.device_descriptor` and pass `self.device_id` where we were passing the descriptor.

armills replied:
Also, to clean up the existing code, let's remove the prefix "KeyboardRemote: " from all of these log messages. Logs are already identified by their platform.

MrMep commented:
Thanks!
Documentation updated.

emlove reviewed (commented):
Looks good. I've just included some minor feedback.

MrMep committed:
Update keyboard_remote.py

armills commented on homeassistant/components/keyboard_remote.py:
```diff
@@ -67,22 +47,18 @@
+    if not config.get(DEVICE_DESCRIPTOR) and\
+       not config.get(DEVICE_NAME):
+        _LOGGER.debug(
+            'Error: No device_descriptor\
```
Just one more, I missed this last time: Remove "Error: " and make this function `_LOGGER.error`.

MrMep committed:
Update keyboard_remote.py

emlove committed:
Unwrap logger error

armills commented:
Couldn't bear to bug you again. :) Thanks for being so responsive.

MrMep commented:
Thanks to you!
I think I did everything you asked for.

armills commented:
Looks good! Nice new feature and a good cleanup of the existing logic. 🎉
